ruby/bps/lib/bps.rb

```diff
@@ -2,75 +2,53 @@
 require 'cgi'
 
 module BPS
+  autoload :Coercer, 'bps/coercer'
+
   module Publisher
-    class Abstract
-      # def initialize; end
+    autoload :Abstract, 'bps/publisher/abstract'
 
-      def topic(_name)
-        raise 'not implemented'
+    def self.register(*schemes, &resolver)
+      @registry ||= {}
+      schemes.each do |scheme|
+        @registry[scheme] = resolver
       end
-
-      def close; end
     end
 
-    module Topic
-      class Abstract
-        # def initialize; end
+    def self.resolve(url, coercer: nil)
+      url = url.is_a?(::URI) ? url.dup : URI.parse(url)
+      rsl = @registry[url.scheme]
+      raise ArgumentError, "Unable to resolve publisher #{url}, scheme #{url.scheme} is not registered" unless rsl
 
-        def publish(_msg_data, **_opts)
-          raise 'not implemented'
-        end
+      opts = {}
+      CGI.parse(url.query.to_s).each do |key, values|
+        opts[key.to_sym] = values.first
       end
+      opts = coercer.coercer(opts) if coercer
```

Comment:
Parsing options from query params as `string -> string` is not great as it breaks the actual kafka library and while casting manually is a chore. For now, I am allowing custom coercion but it's only a start ...

Reply:
not cosmetics, bug :)